README.rst
pytest-rerunfailures
====================

pytest-rerunfailures is a plugin for `pytest <https://pytest.org>`_ that
re-runs tests to eliminate intermittent failures.

.. image:: https://img.shields.io/badge/license-MPL%202.0-blue.svg
   :target: https://github.com/pytest-dev/pytest-rerunfailures/blob/master/LICENSE
   :alt: License
.. image:: https://img.shields.io/pypi/v/pytest-rerunfailures.svg
   :target: https://pypi.python.org/pypi/pytest-rerunfailures/
   :alt: PyPI
.. image:: https://github.com/pytest-dev/pytest-rerunfailures/workflows/Test/badge.svg
   :target: https://github.com/pytest-dev/pytest-rerunfailures/actions
   :alt: GitHub Actions

Requirements
------------

You will need the following prerequisites in order to use pytest-rerunfailures:

- Python 3.6, up to 3.10, or PyPy3
- pytest 5.3 or newer

This package is currently tested against the last 5 minor pytest releases. In
case you work with an older version of pytest you should consider updating or
use one of the earlier versions of this package.

Installation
------------

To install pytest-rerunfailures:

.. code-block:: bash

  $ pip install pytest-rerunfailures

Re-run all failures
-------------------

To re-run all test failures, use the ``--reruns`` command line option with the
maximum number of times you'd like the tests to run:

.. code-block:: bash

  $ pytest --reruns 5

Failed fixture or setup_class will also be re-executed.

To add a delay time between re-runs use the ``--reruns-delay`` command line
option with the amount of seconds that you would like wait before the next
test re-run is launched:

.. code-block:: bash

   $ pytest --reruns 5 --reruns-delay 1

To use a random backoff to vary the delay, use the ``--reruns-delay-backoff``
command line options. This causes the delay to be picked randomly between 0
and a maximum value that increases on every failure (starting value is set by
``--reruns-delay``):

.. code-block:: bash

   $ pytest --reruns 5 --reruns-delay 1 --reruns-delay-backoff


Re-run all failures matching certain expressions
------------------------------------------------

To re-run only those failures that match a certain list of expressions, use the
``--only-rerun`` flag and pass it a regular expression. For example,
the following would only rerun those errors that match ``AssertionError``:

.. code-block:: bash

   $ pytest --reruns 5 --only-rerun AssertionError

Passing the flag multiple times accumulates the arguments, so the following
would only rerun those errors that match ``AssertionError`` or ``ValueError``:

.. code-block:: bash

   $ pytest --reruns 5 --only-rerun AssertionError --only-rerun ValueError

Re-run individual failures
--------------------------

To mark individual tests as flaky, and have them automatically re-run when they
fail, add the ``flaky`` mark with the maximum number of times you'd like the
test to run:

.. code-block:: python

  @pytest.mark.flaky(reruns=5)
  def test_example():
      import random
      assert random.choice([True, False])

Note that when teardown fails, two reports are generated for the case, one for
the test case and the other for the teardown error.

You can also specify the delay time and backoff in the marker:

.. code-block:: python

  @pytest.mark.flaky(reruns=5, reruns_delay=2, reruns_delay_backoff=False)
  def test_example():
      import random
      assert random.choice([True, False])

You can also specify an optional ``condition`` in the re-run marker:

.. code-block:: python

   @pytest.mark.flaky(reruns=5, condition=sys.platform.startswith("win32"))
   def test_example():
      import random
      assert random.choice([True, False])

You can use ``@pytest.mark.flaky(condition)`` similarly as ``@pytest.mark.skipif(condition)``, see `pytest-mark-skipif <https://docs.pytest.org/en/6.2.x/reference.html#pytest-mark-skipif>`_

.. code-block:: python

    @pytest.mark.flaky(reruns=2,condition="sys.platform.startswith('win32')")
    def test_example():
        import random
        assert random.choice([True, False])
    # totally same as the above
    @pytest.mark.flaky(reruns=2,condition=sys.platform.startswith("win32"))
    def test_example():
      import random
      assert random.choice([True, False])

Note that the test will re-run for any ``condition`` that is truthy.

Output
------

Here's an example of the output provided by the plugin when run with
``--reruns 2`` and ``-r aR``::

  test_report.py RRF

  ================================== FAILURES ==================================
  __________________________________ test_fail _________________________________

      def test_fail():
  >       assert False
  E       assert False

  test_report.py:9: AssertionError
  ============================ rerun test summary info =========================
  RERUN test_report.py::test_fail
  RERUN test_report.py::test_fail
  ============================ short test summary info =========================
  FAIL test_report.py::test_fail
  ======================= 1 failed, 2 rerun in 0.02 seconds ====================

Note that output will show all re-runs. Tests that fail on all the re-runs will
be marked as failed.

Compatibility
-------------

* This plugin may *not* be used with class, module, and package level fixtures.
* This plugin is *not* compatible with pytest-xdist's --looponfail flag.
* This plugin is *not* compatible with the core --pdb flag.

Resources
---------

- `Issue Tracker <https://github.com/pytest-dev/pytest-rerunfailures/issues>`_
- `Code <https://github.com/pytest-dev/pytest-rerunfailures/>`_

Development
-----------

* Test execution count can be retrieved from the ``execution_count`` attribute
  in test ``item``'s object. Example:

  .. code-block:: python

    @hookimpl(tryfirst=True)
    def pytest_runtest_makereport(item, call):
        print(item.execution_count)


CHANGES.rst
Changelog
=========

10.2 (2021-09-17)
-----------------

- Add support for Python 3.10 (as of Python 3.10.rc2).
  (Thanks to `@hugovk <https://github.com/hugovk>`_ for the PR.)


10.1 (2021-07-02)
-----------------

Features
++++++++

- Allows using a ``str`` as condition for
  ``@pytest.mark.flaky(condition)``
  which gets evaluated dynamically similarly to
  ``@pytest.mark.skipif(condition)``.
  (`#162 <https://github.com/pytest-dev/pytest-rerunfailures/pull/162>`_
  provided by `@15klli <https://github.com/15klli>`_)

10.0 (2021-05-26)
-----------------

Backwards incompatible changes
++++++++++++++++++++++++++++++

- Drop support for Python 3.5.

- Drop support for pytest < 5.3.

Features
++++++++

- Add ``condition`` keyword argument to the re-run marker.
  (Thanks to `@BeyondEvil`_ for the PR.)

- Add support for Python 3.9.
  (Thanks to `@digitronik`_ for the PR.)

- Add support for pytest 6.3.
  (Thanks to `@bluetech`_ for the PR.)

- Add compatibility with ``pytest-xdist >= 2.0``.
  (Thanks to `@bluetech`_ for the PR.)

Other changes
+++++++++++++

- Check for the resultlog by feature and not by version as pytest master does
  not provide a consistent version.

.. _@BeyondEvil: https://github.com/BeyondEvil
.. _@digitronik: https://github.com/digitronik
.. _@bluetech: https://github.com/bluetech

9.1.1 (2020-09-29)
------------------

Compatibility fix.
++++++++++++++++++

- Ignore ``--result-log`` command line option when used together with ``pytest
  >= 6.1.0``, as it was removed there. This is a quick fix, use an older
  version of pytest, if you want to keep this feature for now.
  (Thanks to `@ntessore`_ for the PR)

- Support up to pytest 6.1.0.

.. _@ntessore: https://github.com/ntessore


9.1 (2020-08-26)
----------------

Features
++++++++

- Add a new flag ``--only-rerun`` to allow for users to rerun only certain
  errors.

Other changes
+++++++++++++

- Drop dependency on ``mock``.

- Add support for pre-commit and add a linting tox target.
  (`#117 <https://github.com/pytest-dev/pytest-rerunfailures/pull/117>`_)
  (PR from `@gnikonorov`_)

.. _@gnikonorov: https://github.com/gnikonorov


9.0 (2020-03-18)
----------------

Backwards incompatible changes
++++++++++++++++++++++++++++++

- Drop support for pytest version 4.4, 4.5 and 4.6.

- Drop support for Python 2.7.


Features
++++++++

- Add support for pytest 5.4.

- Add support for Python 3.8.


8.0 (2019-11-18)
----------------

Backwards incompatible changes
++++++++++++++++++++++++++++++

- Drop support for pytest version 3.10, 4.0, 4.1, 4.2 and 4.3

- Drop support for Python 3.4.

Features
++++++++

- Add support for pytest version 4.4, 4.5, 4.6, 5.0, 5.1 and 5.2.

Bug fixes
+++++++++

- Explicitly depend on setuptools to ensure installation when working in
  environments without it.
  (`#98 <https://github.com/pytest-dev/pytest-rerunfailures/pull/98>`_)
  (PR from `@Eric-Arellano`_)

.. _@Eric-Arellano: https://github.com/Eric-Arellano


7.0 (2019-03-28)
----------------

Backwards incompatible changes
++++++++++++++++++++++++++++++

- Drop support for pytest version 3.8 and 3.9.

Features
++++++++

- Add support for pytest version 4.2 and 4.3.

Bug fixes
+++++++++

- Fixed #83 issue about ignored ``pytest_runtest_logfinish`` hooks.
  (`#83 <https://github.com/pytest-dev/pytest-rerunfailures/issues/83>`_)
  (PR from `@KillAChicken`_)

.. _@KillAChicken: https://github.com/KillAChicken


6.0 (2019-01-08)
----------------

Backwards incompatible changes
++++++++++++++++++++++++++++++

- Drop support for pytest version 3.6 and 3.7.

Features
++++++++

- Add support for pytest version 4.0 and 4.1.

Bug fixes
+++++++++

- Fixed #77 regression issue introduced in 4.2 related to the ``rerun``
  attribute on the test report.
  (`#77 <https://github.com/pytest-dev/pytest-rerunfailures/issues/77>`_)
  (Thanks to `@RibeiroAna`_ for the PR).

.. _@RibeiroAna: https://github.com/RibeiroAna


5.0 (2018-11-06)
----------------

- Drop support for pytest versions < 3.6 to reduce the maintenance burden.

- Add support up to pytest version 3.10. Thus supporting the newest 5 pytest
  releases.

- Add support for Python 3.7.

- Fix issue can occur when used together with `pytest-flake8`
  (`#73 <https://github.com/pytest-dev/pytest-rerunfailures/issues/73>`_)


4.2 (2018-10-04)
----------------

- Fixed #64 issue related to ``setup_class`` and ``fixture`` executions on
  rerun (Thanks to `@OlegKuzovkov`_ for the PR).

- Added new ``execution_count`` attribute to reflect the number of test case
  executions according to #67 issue. (Thanks to `@OlegKuzovkov`_ for the PR).

.. _@OlegKuzovkov: https://github.com/OlegKuzovkov


4.1 (2018-05-23)
----------------

- Add support for pytest 3.6 by using ``Node.get_closest_marker()`` (Thanks to
  `@The-Compiler`_ for the PR).

.. _@The-Compiler: https://github.com/The-Compiler

4.0 (2017-12-23)
----------------

- Added option to add a delay time between test re-runs (Thanks to `@Kanguros`_
  for the PR).

- Added support for pytest >= 3.3.

- Drop support for pytest < 2.8.7.

.. _@Kanguros: https://github.com/Kanguros


3.1 (2017-08-29)
----------------

- Restored compatibility with pytest-xdist. (Thanks to `@davehunt`_ for the PR)

.. _@davehunt: https://github.com/davehunt


3.0 (2017-08-17)
----------------

- Add support for Python 3.6.

- Add support for pytest 2.9 up to 3.2

- Drop support for Python 2.6 and 3.3.

- Drop support for pytest < 2.7.


2.2 (2017-06-23)
----------------

- Ensure that other plugins can run after this one, in case of a global setting
  ``--rerun=0``. (Thanks to `@sublee`_ for the PR)

.. _@sublee: https://github.com/sublee


pytest_rerunfailures.py
import os
import platform
import random
import re
import sys
import time
import traceback
import warnings

import pytest
from _pytest.outcomes import fail
from _pytest.runner import runtestprotocol
from pkg_resources import DistributionNotFound
from pkg_resources import get_distribution
from pkg_resources import parse_version

HAS_RESULTLOG = False
DELAY_BACKOFF_BASE_DEFAULT = 2

try:
    from _pytest.resultlog import ResultLog

    HAS_RESULTLOG = True
except ImportError:
    # We have a pytest >= 6.1
    pass


PYTEST_GTE_54 = parse_version(pytest.__version__) >= parse_version("5.4")
PYTEST_GTE_63 = parse_version(pytest.__version__) >= parse_version("6.3.0.dev")


def works_with_current_xdist():
    """Return compatibility with installed pytest-xdist version.

    When running tests in parallel using pytest-xdist < 1.20.0, the first
    report that is logged will finish and terminate the current node rather
    rerunning the test. Thus we must skip logging of intermediate results under
    these circumstances, otherwise no test is rerun.

    """
    try:
        d = get_distribution("pytest-xdist")
        return d.parsed_version >= parse_version("1.20")
    except DistributionNotFound:
        return None


# command line options
def pytest_addoption(parser):
    group = parser.getgroup(
        "rerunfailures", "re-run failing tests to eliminate flaky failures"
    )
    group._addoption(
        "--only-rerun",
        action="append",
        dest="only_rerun",
        type=str,
        default=None,
        help="If passed, only rerun errors matching the regex provided. "
        "Pass this flag multiple times to accumulate a list of regexes "
        "to match",
    )
    group._addoption(
        "--reruns",
        action="store",
        dest="reruns",
        type=int,
        default=0,
        help="number of times to re-run failed tests. defaults to 0.",
    )
    group._addoption(
        "--reruns-delay",
        action="store",
        dest="reruns_delay",
        type=float,
        default=0,
        help="add time (seconds) delay between reruns.",
    )
    group._addoption(
        "--reruns-delay-backoff",
        action="store_true",
        dest="reruns_delay_backoff",
        help="vary delay between reruns with random exponential backoff "
             "algorithm. Max delay between reruns starts from 'reruns-delay'."
    )
    group._addoption(
        "--reruns-delay-backoff-base",
        dest="reruns_delay_backoff_base",
        type=int,
        default=DELAY_BACKOFF_BASE_DEFAULT,
        help="Set the base for the exponentiation operation that computes the "
             "upper bound for the random backoff delay."
    )
    group._addoption(
        "--reruns-delay-backoff-max",
        dest="reruns_delay_backoff_max",
        type=int,
        help="The maximum value (seconds) used for the upper bound in "
             "selecting the random backoff delay."
    )


def _get_resultlog(config):
    if not HAS_RESULTLOG:
        return None
    elif PYTEST_GTE_54:
        # hack
        from _pytest.resultlog import resultlog_key

        return config._store.get(resultlog_key, default=None)
    else:
        return getattr(config, "_resultlog", None)


def _set_resultlog(config, resultlog):
    if not HAS_RESULTLOG:
        pass
    elif PYTEST_GTE_54:
        # hack
        from _pytest.resultlog import resultlog_key

        config._store[resultlog_key] = resultlog
    else:
        config._resultlog = resultlog


# making sure the options make sense
# should run before / at the beginning of pytest_cmdline_main
def check_options(config):
    val = config.getvalue
    if not val("collectonly"):
        if config.option.reruns != 0:
            if config.option.usepdb:  # a core option
                raise pytest.UsageError("--reruns incompatible with --pdb")

    resultlog = _get_resultlog(config)
    if resultlog:
        logfile = resultlog.logfile
        config.pluginmanager.unregister(resultlog)
        new_resultlog = RerunResultLog(config, logfile)
        _set_resultlog(config, new_resultlog)
        config.pluginmanager.register(new_resultlog)


def _get_marker(item):
    try:
        return item.get_closest_marker("flaky")
    except AttributeError:
        # pytest < 3.6
        return item.get_marker("flaky")


# Class to generate the backoff delay given the number of attempts.
class RerunDelay():
    def __init__(self, delay):
        self._delay = delay

        if self._delay < 0:
            self._delay = 0
            warnings.warn(
                "Delay time between re-runs cannot be < 0. Using default "
                "value: 0"
            )

    def next_rerun_delay(self, iteration_nr):
        raise NotImplementedError


# Constant delay among reruns.
class RerunDelayConst(RerunDelay):
    def next_rerun_delay(self, _):
        return self._delay


# Random exponential backoff delay.
class RerunDelayExp(RerunDelay):
    def __init__(self, factor, base, max):
        super().__init__(factor)
        self._base = base
        self._max = max

        if self._base < 0:
            self._base = DELAY_BACKOFF_BASE_DEFAULT
            warnings.warn(
                "Exponential backoff base cannot be < 0. Using default value "
                f"{DELAY_BACKOFF_BASE_DEFAULT}"
            )

        if self._max and self._max < 0:
            self._max = None
            warnings.warn(
                "Max value for exponential backoff cannot be < 0. "
                "Not enforcing it."
            )

    def next_rerun_delay(self, iteration_nr):
        upper = self._delay * (self._base**iteration_nr)
        if self._max:
            upper = min(self._max, upper)

        return random.randint(self._delay, upper)


def get_reruns_condition(item):
    rerun_marker = _get_marker(item)

    condition = True
    if rerun_marker is not None and "condition" in rerun_marker.kwargs:
        condition = evaluate_condition(
            item, rerun_marker, rerun_marker.kwargs["condition"]
        )

    return condition


def evaluate_condition(item, mark, condition: object) -> bool:
    # copy from python3.8 _pytest.skipping.py

    result = False
    # String condition.
    if isinstance(condition, str):
        globals_ = {
            "os": os,
            "sys": sys,
            "platform": platform,
            "config": item.config,
        }
        if hasattr(item, "obj"):
            globals_.update(item.obj.__globals__)  # type: ignore[attr-defined]
        try:
            filename = f"<{mark.name} condition>"
            condition_code = compile(condition, filename, "eval")
            result = eval(condition_code, globals_)
        except SyntaxError as exc:
            msglines = [
                "Error evaluating %r condition" % mark.name,
                "    " + condition,
                "    " + " " * (exc.offset or 0) + "^",
                "SyntaxError: invalid syntax",
            ]
            fail("\n".join(msglines), pytrace=False)
        except Exception as exc:
            msglines = [
                "Error evaluating %r condition" % mark.name,
                "    " + condition,
                *traceback.format_exception_only(type(exc), exc),
            ]
            fail("\n".join(msglines), pytrace=False)

    # Boolean condition.
    else:
        try:
            result = bool(condition)
        except Exception as exc:
            msglines = [
                "Error evaluating %r condition as a boolean" % mark.name,
                *traceback.format_exception_only(type(exc), exc),
            ]
            fail("\n".join(msglines), pytrace=False)
    return result


class RerunManager():
    """
    Manager for rerunning failed tests.
    """

    def __init__(self, item, reruns=None, delay=None, delay_backoff=None,
                 delay_backoff_base=None, delay_backoff_max=None):
        self._item = item

        self._reruns = reruns or 0

        delay = delay or 0
        self._delay_generator = RerunDelayConst(delay)
        if delay_backoff:
            self._delay_generator = RerunDelayExp(
                delay,
                delay_backoff_base or DELAY_BACKOFF_BASE_DEFAULT,
                delay_backoff_max
            )

        self._execution_count = 0

    @property
    def max_reruns(self):
        return self._reruns

    def match_rerun_policy(self, report):
        """
        Check if @report matches the rerun policy. Return a string with the
        reason if matching, or None otherwise
        """
        assert report.outcome == "failed"

        raise NotImplementedError()

    def next_delay(self):
        return self._delay_generator.next_rerun_delay(self._execution_count)

    def prepare_rerun_test(self, report):
        self._execution_count += 1

        report.outcome = "rerun"
        time.sleep(
            self._delay_generator.next_rerun_delay(self._execution_count - 1)
        )

        if not (hasattr(self._item.config, "workerinput") or
                hasattr(self._item.config, "slaveinput")) or \
           works_with_current_xdist():
            # will rerun test, log intermediate result
            self._item.ihook.pytest_runtest_logreport(report=report)

        # cleaning item's cashed results from any level of setups
        _remove_cached_results_from_failed_fixtures(self._item)
        _remove_failed_setup_state_from_session(self._item)


class FlakyTestRerunManager(RerunManager):
    """
    per-test rerunning setting.
    """

    def __init__(self, item):
        reruns = None
        delay = None
        delay_backoff = None
        delay_backoff_base = None
        delay_backoff_max = None
        self._rerun_marker = _get_marker(item)
        if self._rerun_marker:
            reruns = 1
            if "reruns" in self._rerun_marker.kwargs:
                # check for keyword arguments
                reruns = self._rerun_marker.kwargs["reruns"]
            elif len(self._rerun_marker.args) > 0:
                # check for arguments
                reruns = self._rerun_marker.args[0]

            if "reruns_delay" in self._rerun_marker.kwargs:
                delay = self._rerun_marker.kwargs["reruns_delay"]
            elif len(self._rerun_marker.args) > 1:
                # check for argument
                delay = self._rerun_marker.args[1]

            if "reruns_delay_backoff" in self._rerun_marker.kwargs:
                delay_backoff = self._rerun_marker.kwargs["reruns_delay_backoff"]
            elif len(self._rerun_marker.args) > 2:
                # check for argument
                delay_backoff = self._rerun_marker.args[2]

            if "reruns_delay_backoff_base" in self._rerun_marker.kwargs:
                delay_backoff_base = \
                    self._rerun_marker.kwargs["reruns_delay_backoff_base"]
            elif len(self._rerun_marker.args) > 3:
                # check for argument
                delay_backoff_base = self._rerun_marker.args[3]

            if "reruns_delay_backoff_max" in self._rerun_marker.kwargs:
                delay_backoff_max = \
                    self._rerun_marker.kwargs["reruns_delay_backoff_max"]
            elif len(self._rerun_marker.args) > 4:
                # check for argument
                delay_backoff_max = self._rerun_marker.args[4]

        super().__init__(
            item,
            reruns,
            delay,
            delay_backoff,
            delay_backoff_base,
            delay_backoff_max
        )

    def evaluate_rerun_policy(self, report):
        assert report.outcome == "failed"

        if self._execution_count + 1 > self.max_reruns:
            return False

        condition = get_reruns_condition(self._item)
        if condition:
            report.reason = ""
            return True
        return False


class GlobalRerunManager(RerunManager):
    """
    Global test rerunning policy
    """
    def __init__(self, item):
        delay = item.session.config.option.reruns_delay
        delay_backoff = item.session.config.option.reruns_delay_backoff
        delay_backoff_base = \
            item.session.config.option.reruns_delay_backoff_base
        delay_backoff_max = item.session.config.option.reruns_delay_backoff_max

        self._rerun_errors = item.session.config.option.only_rerun

        super().__init__(
            item,
            item.session.config.option.reruns,
            delay,
            delay_backoff,
            delay_backoff_base,
            delay_backoff_max
        )

    def evaluate_rerun_policy(self, report):
        assert report.outcome == "failed"

        if self._execution_count + 1 > self.max_reruns:
            return False

        if not self._rerun_errors:
            report.reason = ""
            return True

        for rerun_regex in self._rerun_errors:
            if re.search(rerun_regex, report.longrepr.reprcrash.message):
                report.reason = f"Error matching '{rerun_regex}'"
                return True
        return False


def _remove_cached_results_from_failed_fixtures(item):
    """Note: remove all cached_result attribute from every fixture."""
    cached_result = "cached_result"
    fixture_info = getattr(item, "_fixtureinfo", None)
    for fixture_def_str in getattr(fixture_info, "name2fixturedefs", ()):
        fixture_defs = fixture_info.name2fixturedefs[fixture_def_str]
        for fixture_def in fixture_defs:
            if getattr(fixture_def, cached_result, None) is not None:
                result, _, err = getattr(fixture_def, cached_result)
                if err:  # Deleting cached results for only failed fixtures
                    if PYTEST_GTE_54:
                        setattr(fixture_def, cached_result, None)
                    else:
                        delattr(fixture_def, cached_result)


def _remove_failed_setup_state_from_session(item):
    """
    Clean up setup state.

    Note: remove all failures from every node in _setupstate stack
          and clean the stack itself
    """
    setup_state = item.session._setupstate
    if PYTEST_GTE_63:
        setup_state.stack = {}
    else:
        for node in setup_state.stack:
            if hasattr(node, "_prepare_exc"):
                del node._prepare_exc
        setup_state.stack = []


def _failed_test(report):
    """
    Return whether the test has failed according to the @report. Failed means
    the test didn't pass or xfailed.
    """
    xfail = hasattr(report, "wasxfail")
    return (
        report.outcome == "failed" and
        report.failed and
        not xfail
    )


def pytest_configure(config):
    # add flaky marker
    config.addinivalue_line(
        "markers",
        "flaky(reruns=1, reruns_delay=0, reruns_delay_backoff=True, "
        "reruns_delay_backoff_base=2, reruns_delay_backoff_max=3600): mark "
        "test to re-run up to 'reruns' times. Add a delay of 'reruns_delay' "
        "seconds between re-runs. Vary delay according to random exponential "
        "backoff with 'reruns_delay_backoff' (base and max delay set with "
        "'reruns_delay_backoff_base' and 'reruns_delay_backoff_max' "
        "arguments)",
    )


def pytest_runtest_protocol(item, nextitem):
    """
    Run the test protocol.

    Note: when teardown fails, two reports are generated for the case, one for
    the test case and the other for the teardown error.
    """

    test_rerun_policy = FlakyTestRerunManager(item)
    global_rerun_policy = GlobalRerunManager(item)

    if not test_rerun_policy.max_reruns and not global_rerun_policy.max_reruns:
        # global setting is not specified, and this test is not marked with
        # flaky
        return

    # while this doesn't need to be run with every item, it will fail on the
    # first item if necessary
    check_options(item.session.config)
    item.execution_count = 0

    need_to_run = True
    while need_to_run:
        item.execution_count += 1
        item.ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
        reports = runtestprotocol(item, nextitem=nextitem, log=False)

        for report in reports:  # 3 reports: setup, call, teardown
            report.rerun = item.execution_count - 1
            if _failed_test(report):
                if global_rerun_policy.evaluate_rerun_policy(report):
                    global_rerun_policy.prepare_rerun_test(report)
                    break  # trigger rerun
                elif test_rerun_policy.evaluate_rerun_policy(report):
                    # Reset @global_rerun_policy execution count.
                    global_rerun_policy = GlobalRerunManager(item)
                    test_rerun_policy.prepare_rerun_test(report)
                    break  # trigger rerun

            item.ihook.pytest_runtest_logreport(report=report)
        else:
            need_to_run = False

        item.ihook.pytest_runtest_logfinish(nodeid=item.nodeid, location=item.location)

    return True


def pytest_report_teststatus(report):
    # Adapted from https://pytest.org/latest/_modules/_pytest/skipping.html
    if report.outcome == "rerun":
        rerun_string = "RERUN"
        if report.reason:
            rerun_string += f": {report.reason}"
        return "rerun", "R", (rerun_string, {"yellow": True})


def pytest_terminal_summary(terminalreporter):
    # Adapted from https://pytest.org/latest/_modules/_pytest/skipping.html
    tr = terminalreporter
    if not tr.reportchars:
        return

    lines = []
    for char in tr.reportchars:
        if char in "rR":
            show_rerun(terminalreporter, lines)

    if lines:
        tr._tw.sep("=", "rerun test summary info")
        for line in lines:
            tr._tw.line(line)


def show_rerun(terminalreporter, lines):
    rerun = terminalreporter.stats.get("rerun")
    if rerun:
        for rep in rerun:
            pos = rep.nodeid
            lines.append(f"RERUN {pos}")


if HAS_RESULTLOG:

    class RerunResultLog(ResultLog):
        def __init__(self, config, logfile):
            ResultLog.__init__(self, config, logfile)

        def pytest_runtest_logreport(self, report):
            """Add support for rerun report."""
            if report.when != "call" and report.passed:
                return
            res = self.config.hook.pytest_report_teststatus(report=report)
            code = res[1]
            if code == "x":
                longrepr = str(report.longrepr)
            elif code == "X":
                longrepr = ""
            elif report.passed:
                longrepr = ""
            elif report.failed:
                longrepr = str(report.longrepr)
            elif report.skipped:
                longrepr = str(report.longrepr[2])
            elif report.outcome == "rerun":
                longrepr = str(report.longrepr)
            else:
                longrepr = str(report.longrepr)

            self.log_outcome(report, code, longrepr)
